src/swap/v2/router/swap-router.ts

```diff
@@ -0,0 +1,203 @@
+export async function generateSwapRouterTxns({
+  initiatorAddr,
+  client,
+  network,
+  swapType,
+  route
+}: GenerateSwapRouterTxnsParams) {
+  const suggestedParams = await client.getTransactionParams().do();
+
+  const [assetInID, intermediaryAssetID, assetOutID] = [
+    getAssetId(route[0].quote.amount_in.asset),
+    getAssetId(route[0].quote.amount_out.asset),
+    getAssetId(route[1].quote.amount_out.asset)
+  ];
+  const [assetInAmount, assetOutAmount] = [
+    Number(getAssetInFromSwapRoute(route).amount),
+    Number(getAssetOutFromSwapRoute(route).amount)
+  ];
+  const [pool1Address, pool2Address] = [route[0].pool.address, route[1].pool.address];
+
+  const isAssetInAlgo = isAlgo(assetInID);
+
+  const routerAppID = getSwapRouterAppID(network);
+
+  const inputTxn = isAssetInAlgo
+    ? algosdk.makePaymentTxnWithSuggestedParamsFromObject({
+        from: initiatorAddr,
+        to: getApplicationAddress(routerAppID),
+        amount: assetInAmount,
+        suggestedParams
+      })
+    : algosdk.makeAssetTransferTxnWithSuggestedParamsFromObject({
+        from: initiatorAddr,
+        to: getApplicationAddress(routerAppID),
+        amount: assetInAmount,
+        assetIndex: assetInID,
+        suggestedParams
+      });
+
+  const routerAppCallTxn = algosdk.makeApplicationNoOpTxnFromObject({
+    from: initiatorAddr,
+    appIndex: routerAppID,
+    appArgs: [
+      V2_SWAP_APP_CALL_ARG_ENCODED,
+      V2_SWAP_APP_CALL_SWAP_TYPE_ARGS_ENCODED[swapType],
+      algosdk.encodeUint64(assetOutAmount)
+    ],
+    foreignApps: [getValidatorAppID(network, CONTRACT_VERSION.V2)],
+    foreignAssets: [assetInID, intermediaryAssetID, assetOutID],
+    accounts: [pool1Address, pool2Address],
+    suggestedParams
+  });
+
+  routerAppCallTxn.fee =
+    ALGORAND_MIN_TX_FEE * (SWAP_ROUTER_INNER_TXN_COUNT[swapType] + 1);
+
+  const txnList = [inputTxn, routerAppCallTxn];
+
+  const optInRequiredAssetIDs = await getSwapRouterAppOptInRequiredAssetIDs({
+    client,
+    routerAppID,
+    assetIDs: [assetInID, intermediaryAssetID, assetOutID]
+  });
+
+  if (optInRequiredAssetIDs.length > 0) {
+    const routerAppAssetOptInTxn = await generateSwapRouterAssetOptInTransaction({
+      client,
+      accountAddress: initiatorAddr,
+      assetIDs: optInRequiredAssetIDs,
+      routerAppID
+    });
+
+    txnList.unshift(routerAppAssetOptInTxn);
+  }
+
+  const txGroup = algosdk.assignGroupID(txnList);
+
+  return txGroup.map((txn: Transaction) => ({
+    txn,
+    signers: [initiatorAddr]
+  }));
+}
+
+export async function getSwapRouterAppOptInRequiredAssetIDs({
+  client,
+  routerAppID,
+  assetIDs
+}: {
+  client: AlgodClient;
+  routerAppID: number;
+  assetIDs: number[];
+}) {
+  const swapRouterAppAddress = getApplicationAddress(routerAppID);
+  const accountInfo = await client.accountInformation(swapRouterAppAddress).do();
+  const appOptedInAssetIDs = accountInfo.assets.map(
+    (asset: {"asset-id": number}) => asset["asset-id"]
+  );
```

Comment:
I think we can cast the accountInfo type as AccountInformation and then TSC will properly infer the types for the map below
```suggestion
  const accountInfo = (await client.accountInformation(swapRouterAppAddress).do()) as AccountInformation;
  const appOptedInAssetIDs = accountInfo.assets.map((asset) => asset["asset-id"]);
```